File: packages/adapter-agent-runtime-agently/src/protocol.ts
import {
  validateRuntimeCapabilityContinuation,
  validateRuntimeDriverTurn,
  type RuntimeCapabilityContinuation,
  type RuntimeDriverResult,
  type RuntimeDriverTurn,
  type RuntimeJsonObject,
  type RuntimeJsonValue,
  type RuntimeTaskPackage,
} from "@work-fabric/agent-runtime-spi";

export const AGENTLY_WORKER_PROTOCOL = "workfabric.agent-runtime/1" as const;
export const AGENTLY_WORKER_TURN_PROTOCOL =
  "workfabric.agent-runtime/2" as const;
export const MAX_JSON_DEPTH = 32;
export const MAX_JSON_NODES = 10_000;
export const MAX_JSON_STRING_BYTES = 131_072;

export interface AgentlyWorkerRequestV1 {
  readonly protocol: typeof AGENTLY_WORKER_PROTOCOL;
  readonly command_id: string;
  readonly task: RuntimeTaskPackage;
  readonly provider: { readonly type: "OpenAICompatible"; readonly base_url: string; readonly model: string };
}

export interface AgentlyWorkerRequestV2 {
  readonly protocol: typeof AGENTLY_WORKER_TURN_PROTOCOL;
  readonly command_id: string;
  readonly task: RuntimeTaskPackage;
  readonly continuation: RuntimeCapabilityContinuation | null;
  readonly provider: {
    readonly type: "OpenAICompatible";
    readonly base_url: string;
    readonly model: string;
  };
}

export type AgentlyWorkerRecordV1 =
  | { readonly protocol: typeof AGENTLY_WORKER_PROTOCOL; readonly type: "progress"; readonly command_id: string; readonly sequence: number; readonly progress: number | null; readonly message: string; readonly observed_at: string }
  | { readonly protocol: typeof AGENTLY_WORKER_PROTOCOL; readonly type: "completed"; readonly command_id: string; readonly result: RuntimeDriverResult }
  | { readonly protocol: typeof AGENTLY_WORKER_PROTOCOL; readonly type: "failed"; readonly command_id: string; readonly code: string; readonly message: string; readonly retryable: boolean };

export type AgentlyWorkerTurnRecordV2 =
  | {
      readonly protocol: typeof AGENTLY_WORKER_TURN_PROTOCOL;
      readonly type: "progress";
      readonly command_id: string;
      readonly sequence: number;
      readonly progress: number | null;
      readonly message: string;
      readonly observed_at: string;
    }
  | {
      readonly protocol: typeof AGENTLY_WORKER_TURN_PROTOCOL;
      readonly type: "final" | "capability_request";
      readonly command_id: string;
      readonly turn: RuntimeDriverTurn;
    }
  | {
      readonly protocol: typeof AGENTLY_WORKER_TURN_PROTOCOL;
      readonly type: "failed";
      readonly command_id: string;
      readonly code: string;
      readonly message: string;
      readonly retryable: boolean;
    };

interface JsonBudget { nodes: number; stringBytes: number; readonly seen: WeakSet<object>; }

function ownData(value: object, key: string | symbol, path: string): unknown {
  const descriptor = Object.getOwnPropertyDescriptor(value, key);
  if (descriptor === undefined || !("value" in descriptor)) throw new TypeError(`${path} must be an own data property`);
  return descriptor.value;
}

function plainObject(value: unknown, path: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value) || Object.getPrototypeOf(value) !== Object.prototype) throw new TypeError(`${path} must be a plain object`);
  const keys = Reflect.ownKeys(value);
  const output: Record<string, unknown> = {};
  for (const key of keys) {
    if (typeof key !== "string") throw new TypeError(`${path} contains a symbol key`);
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !("value" in descriptor) || !descriptor.enumerable) throw new TypeError(`${path}.${key} must be an enumerable data property`);
    output[key] = descriptor.value;
  }
  return output;
}

function exactObject(value: unknown, fields: readonly string[], path: string): Record<string, unknown> {
  const object = plainObject(value, path);
  const keys = Object.keys(object);
  if (keys.length !== fields.length || keys.some((key) => !fields.includes(key)) || fields.some((key) => !Object.hasOwn(object, key))) throw new TypeError(`${path} fields are invalid`);
  return object;
}

function countString(value: string, budget: JsonBudget): string {
  budget.stringBytes += Buffer.byteLength(value, "utf8");
  if (budget.stringBytes > MAX_JSON_STRING_BYTES) throw new RangeError("worker JSON strings exceed their bound");
  return value;
}

function json(value: unknown, budget: JsonBudget, depth = 0): RuntimeJsonValue {
  if (depth > MAX_JSON_DEPTH) throw new RangeError("worker JSON exceeds maximum depth");
  budget.nodes += 1;
  if (budget.nodes > MAX_JSON_NODES) throw new RangeError("worker JSON exceeds its node bound");
  if (value === null || typeof value === "boolean") return value;
  if (typeof value === "string") return countString(value, budget);
  if (typeof value === "number") { if (!Number.isFinite(value)) throw new TypeError("worker JSON number is invalid"); return value; }
  if (typeof value !== "object" || value === null) throw new TypeError("worker JSON is invalid");
  if (budget.seen.has(value)) throw new TypeError("worker JSON must not contain references");
  budget.seen.add(value);
  if (Array.isArray(value)) {
    const length = ownData(value, "length", "worker JSON array");
    if (!Number.isSafeInteger(length) || (length as number) < 0) throw new TypeError("worker JSON array length is invalid");
    const keys = Reflect.ownKeys(value);
    if (keys.length !== (length as number) + 1 || keys.some((key) => key !== "length" && (typeof key !== "string" || !/^(0|[1-9][0-9]*)$/.test(key) || Number(key) >= (length as number)))) throw new TypeError("worker JSON array is invalid");
    const output: RuntimeJsonValue[] = [];
    for (let index = 0; index < (length as number); index += 1) output.push(json(ownData(value, String(index), `worker JSON array[${index}]`), budget, depth + 1));
    return output;
  }
  const object = plainObject(value, "worker JSON object");
  const output: Record<string, RuntimeJsonValue> = {};
  for (const [key, child] of Object.entries(object)) {
    if (key.length > 256) throw new RangeError("worker JSON key is too large");
    countString(key, budget);
    output[key] = json(child, budget, depth + 1);
  }
  return output;
}

function string(value: unknown, path: string, maximum: number): string {
  if (typeof value !== "string" || value.length === 0 || value.length > maximum || value.trim() !== value) throw new TypeError(`${path} is invalid`);
  return value;
}

function command(value: unknown, expected: string): void {
  if (value !== expected) throw new TypeError("worker command_id does not match request");
}

function jsonObject(value: unknown, path: string): RuntimeJsonObject {
  const object = plainObject(value, path);
  return object as RuntimeJsonObject;
}

function result(value: unknown): RuntimeDriverResult {
  const parsed = exactObject(value, ["summary", "artifacts", "evidence", "extensions"], "worker result");
  if (!Array.isArray(parsed.summary) || !Array.isArray(parsed.artifacts) || !Array.isArray(parsed.evidence)) throw new TypeError("worker result collections are invalid");
  return {
    summary: parsed.summary.map((item, index) => jsonObject(item, `worker result.summary[${index}]`)),
    artifacts: parsed.artifacts.map((item, index) => jsonObject(item, `worker result.artifacts[${index}]`)),
    evidence: parsed.evidence.map((item, index) => jsonObject(item, `worker result.evidence[${index}]`)),
    extensions: jsonObject(parsed.extensions, "worker result.extensions"),
  };
}

export function parseAgentlyWorkerRecord(value: unknown, expectedCommandId: string): AgentlyWorkerRecordV1 {
  const safe = json(value, { nodes: 0, stringBytes: 0, seen: new WeakSet<object>() });
  const discriminator = plainObject(safe, "worker record");
  if (discriminator.type === "progress") {
    const header = exactObject(discriminator, ["protocol", "type", "command_id", "sequence", "progress", "message", "observed_at"], "worker progress record");
    if (header.protocol !== AGENTLY_WORKER_PROTOCOL) throw new TypeError("worker protocol is unsupported");
    command(header.command_id, expectedCommandId);
    if (!Number.isSafeInteger(header.sequence) || (header.sequence as number) < 1) throw new TypeError("worker progress sequence is invalid");
    if (header.progress !== null && (typeof header.progress !== "number" || !Number.isFinite(header.progress) || header.progress < 0 || header.progress > 1)) throw new TypeError("worker progress value is invalid");
    return { protocol: AGENTLY_WORKER_PROTOCOL, type: "progress", command_id: expectedCommandId, sequence: header.sequence as number, progress: header.progress as number | null, message: string(header.message, "worker progress message", 8_192), observed_at: string(header.observed_at, "worker progress timestamp", 128) };
  }
  if (discriminator.type === "completed") {
    const parsed = exactObject(discriminator, ["protocol", "type", "command_id", "result"], "worker completed record");
    if (parsed.protocol !== AGENTLY_WORKER_PROTOCOL) throw new TypeError("worker protocol is unsupported");
    command(parsed.command_id, expectedCommandId);
    return { protocol: AGENTLY_WORKER_PROTOCOL, type: "completed", command_id: expectedCommandId, result: result(parsed.result) };
  }
  if (discriminator.type === "failed") {
    const parsed = exactObject(discriminator, ["protocol", "type", "command_id", "code", "message", "retryable"], "worker failed record");
    if (parsed.protocol !== AGENTLY_WORKER_PROTOCOL) throw new TypeError("worker protocol is unsupported");
    command(parsed.command_id, expectedCommandId);
    if (typeof parsed.retryable !== "boolean") throw new TypeError("worker retryable is invalid");
    return { protocol: AGENTLY_WORKER_PROTOCOL, type: "failed", command_id: expectedCommandId, code: string(parsed.code, "worker failure code", 128), message: string(parsed.message, "worker failure message", 8_192), retryable: parsed.retryable };
  }
  throw new TypeError("worker record type is unsupported");
}

export function parseAgentlyWorkerTurnRecord(
  value: unknown,
  expectedCommandId: string,
): AgentlyWorkerTurnRecordV2 {
  const safe = json(value, {
    nodes: 0,
    stringBytes: 0,
    seen: new WeakSet<object>(),
  });
  const discriminator = plainObject(safe, "worker turn record");
  if (discriminator.type === "progress") {
    const header = exactObject(
      discriminator,
      [
        "protocol",
        "type",
        "command_id",
        "sequence",
        "progress",
        "message",
        "observed_at",
      ],
      "worker turn progress record",
    );
    if (header.protocol !== AGENTLY_WORKER_TURN_PROTOCOL) {
      throw new TypeError("worker turn protocol is unsupported");
    }
    command(header.command_id, expectedCommandId);
    if (
      !Number.isSafeInteger(header.sequence) ||
      (header.sequence as number) < 1
    ) {
      throw new TypeError("worker progress sequence is invalid");
    }
    if (
      header.progress !== null &&
      (typeof header.progress !== "number" ||
        !Number.isFinite(header.progress) ||
        header.progress < 0 ||
        header.progress > 1)
    ) {
      throw new TypeError("worker progress value is invalid");
    }
    return {
      protocol: AGENTLY_WORKER_TURN_PROTOCOL,
      type: "progress",
      command_id: expectedCommandId,
      sequence: header.sequence as number,
      progress: header.progress as number | null,
      message: string(header.message, "worker progress message", 8_192),
      observed_at: string(
        header.observed_at,
        "worker progress timestamp",
        128,
      ),
    };
  }
  if (discriminator.type === "final") {
    const parsed = exactObject(
      discriminator,
      ["protocol", "type", "command_id", "response"],
      "worker final record",
    );
    if (parsed.protocol !== AGENTLY_WORKER_TURN_PROTOCOL) {
      throw new TypeError("worker turn protocol is unsupported");
    }
    command(parsed.command_id, expectedCommandId);
    return {
      protocol: AGENTLY_WORKER_TURN_PROTOCOL,
      type: "final",
      command_id: expectedCommandId,
      turn: validateRuntimeDriverTurn({
        kind: "final",
        response: parsed.response,
      }),
    };
  }
  if (discriminator.type === "capability_request") {
    const parsed = exactObject(
      discriminator,
      ["protocol", "type", "command_id", "request"],
      "worker capability request record",
    );
    if (parsed.protocol !== AGENTLY_WORKER_TURN_PROTOCOL) {
      throw new TypeError("worker turn protocol is unsupported");
    }
    command(parsed.command_id, expectedCommandId);
    return {
      protocol: AGENTLY_WORKER_TURN_PROTOCOL,
      type: "capability_request",
      command_id: expectedCommandId,
      turn: validateRuntimeDriverTurn({
        kind: "capability_request",
        request: parsed.request,
      }),
    };
  }
  if (discriminator.type === "failed") {
    const parsed = exactObject(
      discriminator,
      [
        "protocol",
        "type",
        "command_id",
        "code",
        "message",
        "retryable",
      ],
      "worker turn failed record",
    );
    if (parsed.protocol !== AGENTLY_WORKER_TURN_PROTOCOL) {
      throw new TypeError("worker turn protocol is unsupported");
    }
    command(parsed.command_id, expectedCommandId);
    if (typeof parsed.retryable !== "boolean") {
      throw new TypeError("worker retryable is invalid");
    }
    return {
      protocol: AGENTLY_WORKER_TURN_PROTOCOL,
      type: "failed",
      command_id: expectedCommandId,
      code: string(parsed.code, "worker failure code", 128),
      message: string(parsed.message, "worker failure message", 8_192),
      retryable: parsed.retryable,
    };
  }
  throw new TypeError("worker turn record type is unsupported");
}

export function normalizeAgentlyWorkerRequestV2(
  value: AgentlyWorkerRequestV2,
): AgentlyWorkerRequestV2 {
  return {
    ...value,
    continuation:
      value.continuation === null
        ? null
        : validateRuntimeCapabilityContinuation(value.continuation),
  };
}

File: runtimes/agently-worker/src/work_fabric_agently_runtime/protocol.py
from __future__ import annotations

import json
import re
import sys
from dataclasses import dataclass
from typing import Literal, Mapping, TextIO, TypeAlias

PROTOCOL = "workfabric.agent-runtime/1"
TURN_PROTOCOL = "workfabric.agent-runtime/2"
MAX_JSON_DEPTH = 32
MAX_JSON_NODES = 10_000
MAX_JSON_STRING_BYTES = 131_072
MAX_JSON_KEY_LENGTH = 256
MAX_STDOUT_LINE_BYTES = 262_144
SECRET_FIELD = re.compile(r"(?:api[_-]?key|secret|token|password|credential)", re.IGNORECASE)

JsonValue: TypeAlias = None | bool | int | float | str | list["JsonValue"] | dict[str, "JsonValue"]


class ProtocolError(ValueError):
    """Raised when an untrusted worker boundary value is invalid."""


@dataclass(frozen=True)
class WorkerRequest:
    protocol: Literal["workfabric.agent-runtime/1", "workfabric.agent-runtime/2"]
    command_id: str
    task: Mapping[str, JsonValue]
    continuation: Mapping[str, JsonValue] | None
    provider_type: Literal["OpenAICompatible"]
    provider_base_url: str
    provider_model: str


@dataclass(frozen=True)
class WorkerRecord:
    protocol: Literal["workfabric.agent-runtime/1", "workfabric.agent-runtime/2"]
    type: Literal["progress", "completed", "final", "capability_request", "failed"]
    command_id: str
    payload: Mapping[str, JsonValue]


@dataclass
class _JsonBudget:
    nodes: int = 0
    string_bytes: int = 0


def _fail(message: str) -> None:
    raise ProtocolError(message)


def usv_string(value: str) -> str:
    return "".join("\ufffd" if 0xD800 <= ord(character) <= 0xDFFF else character for character in value)


def utf16_code_units(value: str) -> int:
    return sum(2 if ord(character) > 0xFFFF else 1 for character in value)


def utf8_usv_bytes(value: str) -> int:
    return len(usv_string(value).encode("utf-8"))


def _string(value: object, label: str, maximum: int) -> str:
    if not isinstance(value, str) or not value or value.strip() != value or utf16_code_units(value) > maximum:
        _fail(f"{label} is invalid")
    return usv_string(value)


def _exact_object(value: object, fields: tuple[str, ...], label: str) -> dict[str, object]:
    if not isinstance(value, dict) or set(value) != set(fields) or len(value) != len(fields):
        _fail(f"{label} contains unknown or missing fields")
    return value


def _json(value: object, budget: _JsonBudget, depth: int = 0) -> JsonValue:
    if depth > MAX_JSON_DEPTH:
        _fail("worker JSON exceeds maximum depth")
    budget.nodes += 1
    if budget.nodes > MAX_JSON_NODES:
        _fail("worker JSON exceeds maximum node bound")
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, str):
        value = usv_string(value)
        budget.string_bytes += utf8_usv_bytes(value)
        if budget.string_bytes > MAX_JSON_STRING_BYTES:
            _fail("worker JSON string bytes exceed maximum bound")
        return value
    if isinstance(value, int) and not isinstance(value, bool):
        return value
    if isinstance(value, float):
        if value != value or value in (float("inf"), float("-inf")):
            _fail("worker JSON number is invalid")
        return value
    if isinstance(value, list):
        return [_json(item, budget, depth + 1) for item in value]
    if not isinstance(value, dict):
        _fail("worker JSON is invalid")
    output: dict[str, JsonValue] = {}
    for key, child in value.items():
        if not isinstance(key, str) or utf16_code_units(key) > MAX_JSON_KEY_LENGTH:
            _fail("worker JSON key is invalid")
        key = usv_string(key)
        if key in output:
            _fail("worker JSON keys collapse under USV normalization")
        budget.string_bytes += utf8_usv_bytes(key)
        if budget.string_bytes > MAX_JSON_STRING_BYTES:
            _fail("worker JSON string bytes exceed maximum bound")
        output[key] = _json(child, budget, depth + 1)
    return output


def _json_object(value: object, label: str) -> dict[str, JsonValue]:
    parsed = _json(value, _JsonBudget())
    if not isinstance(parsed, dict):
        _fail(f"{label} must be a JSON object")
    return parsed


def _reject_secret_fields(value: JsonValue) -> None:
    if isinstance(value, list):
        for item in value:
            _reject_secret_fields(item)
    elif isinstance(value, dict):
        for key, item in value.items():
            if SECRET_FIELD.search(key):
                _fail("task contains a secret-named field")
            _reject_secret_fields(item)


def _validate_role(value: object) -> dict[str, JsonValue]:
    role = _exact_object(value, ("role_id", "version", "display_name", "description", "capability_ids"), "task.role")
    _string(role["role_id"], "task.role.role_id", 128)
    if not isinstance(role["version"], int) or isinstance(role["version"], bool) or role["version"] < 1:
        _fail("task.role.version is invalid")
    _string(role["display_name"], "task.role.display_name", 8_192)
    _string(role["description"], "task.role.description", 8_192)
    capabilities = role["capability_ids"]
    if not isinstance(capabilities, list) or not capabilities:
        _fail("task.role.capability_ids is invalid")
    for item in capabilities:
        _string(item, "task.role.capability_ids", 128)
    return _json_object(role, "task.role")


def _validate_task(value: object) -> dict[str, JsonValue]:
    fields = (
        "tenant_id", "handoff_id", "thread_id", "stream_version", "role", "capability_id",
        "intent", "context_reference", "authority_scope", "acceptance_criteria", "priority",
        "accept_by", "result_due_at", "workspace_path",
    )
    task = _exact_object(value, fields, "task")
    for key in ("tenant_id", "handoff_id", "thread_id", "accept_by", "result_due_at", "workspace_path"):
        _string(task[key], f"task.{key}", 8_192)
    if not isinstance(task["stream_version"], int) or isinstance(task["stream_version"], bool) or task["stream_version"] < 1:
        _fail("task.stream_version is invalid")
    _validate_role(task["role"])
    if task["capability_id"] is not None:
        _string(task["capability_id"], "task.capability_id", 128)
    if task["priority"] not in ("low", "normal", "high", "critical"):
        _fail("task.priority is invalid")
    for key in ("intent", "acceptance_criteria"):
        if not isinstance(task[key], list) or any(not isinstance(item, dict) for item in task[key]):
            _fail(f"task.{key} is invalid")
    if task["context_reference"] is not None and not isinstance(task["context_reference"], dict):
        _fail("task.context_reference is invalid")
    if not isinstance(task["authority_scope"], dict):
        _fail("task.authority_scope is invalid")
    safe_task = _json_object(task, "task")
    _reject_secret_fields(safe_task)
    return safe_task


def _validate_capability_request(value: object) -> dict[str, JsonValue]:
    request = _exact_object(
        value,
        ("invocation_id", "capability_id", "version_constraint", "input", "reason"),
        "continuation.request",
    )
    for field in ("invocation_id", "capability_id", "version_constraint", "reason"):
        _string(request[field], f"continuation.request.{field}", 8_192)
    if not isinstance(request["input"], dict):
        _fail("continuation.request.input is invalid")
    return _json_object(request, "continuation.request")


def _validate_capability_result(value: object) -> dict[str, JsonValue]:
    if not isinstance(value, dict):
        _fail("continuation.result contains unknown or missing fields")
    outcome = value.get("outcome")
    if outcome == "succeeded":
        result = _exact_object(
            value,
            (
                "outcome", "invocation_id", "auxiliary_handoff_id", "candidate",
                "data", "artifacts",
            ),
            "continuation.result",
        )
        candidate = _exact_object(
            result["candidate"],
            (
                "citizen_id", "endpoint_id", "capability_id",
                "capability_version", "contract_digest",
            ),
            "continuation.result.candidate",
        )
        for field in candidate:
            _string(candidate[field], f"continuation.result.candidate.{field}", 256)
        if not isinstance(result["data"], dict):
            _fail("continuation.result.data is invalid")
        if not isinstance(result["artifacts"], list) or any(
            not isinstance(item, dict) for item in result["artifacts"]
        ):
            _fail("continuation.result.artifacts is invalid")
    elif outcome in ("rejected", "failed"):
        result = _exact_object(
            value,
            (
                "outcome", "invocation_id", "auxiliary_handoff_id",
                "code", "message", "retryable",
            ),
            "continuation.result",
        )
        for field in ("invocation_id", "code", "message"):
            _string(result[field], f"continuation.result.{field}", 8_192)
        if result["auxiliary_handoff_id"] is not None:
            _string(
                result["auxiliary_handoff_id"],
                "continuation.result.auxiliary_handoff_id",
                128,
            )
        if not isinstance(result["retryable"], bool):
            _fail("continuation.result.retryable is invalid")
    else:
        _fail("continuation.result.outcome is invalid")
    return _json_object(result, "continuation.result")


def _validate_continuation(value: object) -> dict[str, JsonValue]:
    continuation = _exact_object(
        value,
        ("request", "result"),
        "continuation",
    )
    request = _validate_capability_request(continuation["request"])
    result = _validate_capability_result(continuation["result"])
    if request["invocation_id"] != result["invocation_id"]:
        _fail("continuation invocation_id does not match")
    safe = {"request": request, "result": result}
    _reject_secret_fields(safe)
    return safe


def parse_request(value: object) -> WorkerRequest:
    safe = _json(value, _JsonBudget())
    if not isinstance(safe, dict):
        _fail("request contains unknown or missing fields")
    protocol = safe.get("protocol")
    fields = (
        ("protocol", "command_id", "task", "provider")
        if protocol == PROTOCOL
        else ("protocol", "command_id", "task", "continuation", "provider")
    )
    request = _exact_object(safe, fields, "request")
    if protocol not in (PROTOCOL, TURN_PROTOCOL):
        _fail("request protocol is unsupported")
    command_id = _string(request["command_id"], "request.command_id", 128)
    task = _validate_task(request["task"])
    provider = _exact_object(request["provider"], ("type", "base_url", "model"), "provider")
    if provider["type"] != "OpenAICompatible":
        _fail("provider.type is unsupported")
    return WorkerRequest(
        protocol=protocol,
        command_id=command_id,
        task=task,
        continuation=(
            None
            if protocol == PROTOCOL or request["continuation"] is None
            else _validate_continuation(request["continuation"])
        ),
        provider_type="OpenAICompatible",
        provider_base_url=_string(provider["base_url"], "provider.base_url", 8_192),
        provider_model=_string(provider["model"], "provider.model", 8_192),
    )


def _result(value: object) -> dict[str, JsonValue]:
    result = _exact_object(value, ("summary", "artifacts", "evidence", "extensions"), "result")
    if not isinstance(result["summary"], list) or not result["summary"]:
        _fail("result.summary must be a non-empty array")
    for name in ("summary", "artifacts", "evidence"):
        if not isinstance(result[name], list) or any(not isinstance(item, dict) for item in result[name]):
            _fail(f"result.{name} is invalid")
    if not isinstance(result["extensions"], dict):
        _fail("result.extensions is invalid")
    return _json_object(result, "result")


def progress_record(
    command_id: str,
    sequence: int,
    progress: float | None,
    message: str,
    observed_at: str,
    protocol: Literal["workfabric.agent-runtime/1", "workfabric.agent-runtime/2"] = PROTOCOL,
) -> WorkerRecord:
    if not isinstance(sequence, int) or isinstance(sequence, bool) or sequence < 1:
        _fail("progress sequence is invalid")
    if progress is not None and (not isinstance(progress, (int, float)) or isinstance(progress, bool) or not 0 <= progress <= 1):
        _fail("progress value is invalid")
    if protocol not in (PROTOCOL, TURN_PROTOCOL):
        _fail("record protocol is unsupported")
    return WorkerRecord(protocol, "progress", _string(command_id, "command_id", 128), {
        "sequence": sequence, "progress": progress, "message": _string(message, "progress message", 8_192),
        "observed_at": _string(observed_at, "progress timestamp", 128),
    })


def completed_record(command_id: str, result: object) -> WorkerRecord:
    return WorkerRecord(PROTOCOL, "completed", _string(command_id, "command_id", 128), {"result": _result(result)})


def final_record(command_id: str, response: object) -> WorkerRecord:
    return WorkerRecord(
        TURN_PROTOCOL,
        "final",
        _string(command_id, "command_id", 128),
        {"response": _result(response)},
    )


def capability_request_record(command_id: str, value: object) -> WorkerRecord:
    return WorkerRecord(
        TURN_PROTOCOL,
        "capability_request",
        _string(command_id, "command_id", 128),
        {"request": _validate_capability_request(value)},
    )


def failed_record(
    command_id: str,
    code: str,
    message: str,
    retryable: bool,
    protocol: Literal["workfabric.agent-runtime/1", "workfabric.agent-runtime/2"] = PROTOCOL,
) -> WorkerRecord:
    if not isinstance(retryable, bool):
        _fail("failure retryable is invalid")
    if protocol not in (PROTOCOL, TURN_PROTOCOL):
        _fail("record protocol is unsupported")
    return WorkerRecord(protocol, "failed", _string(command_id, "command_id", 128), {
        "code": _string(code, "failure code", 128), "message": _string(message, "failure message", 8_192), "retryable": retryable,
    })


def record_json(record: WorkerRecord) -> dict[str, JsonValue]:
    return {"protocol": record.protocol, "type": record.type, "command_id": record.command_id, **record.payload}


def write_record(stream: TextIO, record: WorkerRecord) -> None:
    encoded = json.dumps(record_json(record), ensure_ascii=False, separators=(",", ":"), allow_nan=False)
    if len(encoded.encode("utf-8")) > MAX_STDOUT_LINE_BYTES:
        _fail("worker record exceeds stdout bound")
    stream.write(encoded + "\n")
    stream.flush()


def read_request(stream: object = None) -> WorkerRequest:
    input_stream = sys.stdin.buffer if stream is None else stream
    raw = input_stream.read(1_048_577)
    if not isinstance(raw, bytes) or len(raw) > 1_048_576:
        _fail("worker request exceeds input bound")
    try:
        if raw.endswith(b"\n"):
            raw = raw[:-1]
        if not raw or b"\n" in raw or b"\r" in raw:
            _fail("worker request must be exactly one NDJSON record")
        return parse_request(json.loads(raw.decode("utf-8")))
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise ProtocolError("worker request is invalid JSON") from error
